### DISCUSSION TITLE
How would you test file upload ?

### DISCUSSION THREAD

**Discussion by JonathanCabezas** [RobinTail/express-zod-api#2250] [Q&A] [answered]
Hello,

How would you test an endpoint using `ez.upload()` as input ?

I want to do something similar to this [Stack Overflow answer](https://stackoverflow.com/a/57422857/7920723):
```typescript
const fs = require('fs');
const MockExpressRequest = require('mock-express-request');
const FormData = require('form-data');

const form = new FormData();
form.append('my_file',
  fs.createReadStream(path.join(__dirname, 'fixtures', 'file-upload-test.txt'))
);
const request = new MockExpressRequest({
  method: 'POST',
  host: 'localhost',
  url: '/upload',
  headers: form.getHeaders()
});

form.pipe(request);
doUpload(request, response, next);
```

But I don't want to add dependencies just for this test case.

**Comment by RobinTail** [comment, marked as answer]
Hello @JonathanCabezas ,

Upload functionality relies on `express-fileupload` body parser which is not covered by the unit testing helper `testEndpoint()`.
You can use `testEndpoint` with `{ requestProps: { method: "POST", body: data }}` where `data` is already a parsed object, but that would be the test of the endpoint handler, not the parser, like the upload did already happen.

For testing the upload functionality you'd need a system test:
1. Spawn a process running your application: https://github.com/RobinTail/express-zod-api/blob/6e2374873b2ee699079b89717d52ce7594696115/tests/system/example.spec.ts#L19
3. Ensure it's listening (ready to accept requests): https://github.com/RobinTail/express-zod-api/blob/6e2374873b2ee699079b89717d52ce7594696115/tests/system/example.spec.ts#L22
4. Use `FormData` and `fetch` to perform an actual request like this: https://github.com/RobinTail/express-zod-api/blob/6e2374873b2ee699079b89717d52ce7594696115/tests/system/example.spec.ts#L195-L213

**Reply by JonathanCabezas** [reply]
I was able to implement such a test.
Thank you for the quick answer !
